pkg/test/framework/api/component/configuration.go

```diff
@@ -0,0 +1,58 @@
+//  Copyright 2019 Istio Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package component
+
+import "fmt"
+
+var (
+	_ Requirement = &ConfiguredRequirement{}
+)
+
+// Configuration is a marker interface for configuration objects that components take.
+type Configuration fmt.Stringer
+
+// A configured requirement contains the requirement, a configuration object, and the name to give
+// that configured object.
+type ConfiguredRequirement struct {
+	name   string
+	req    Requirement
+	config Configuration
```

Comment:
Might it be the case that `ConfiguredRequirement`s are for a particular variant? For example, if we had a RealPilot and a FakePilot, their configurations may be very different. I wonder if instead of abstractly specifying a `Requirement` if it should be more specifically using `Descriptor`?

Reply:
The idea is that you can use either a Descriptor or an ID here, so if you use a Descriptor it can include the variant. I didn't want to require a Descriptor because the ID is handy, for example if we just want two Echo instances and the test works for both native and kube versions.

I'll add a test case showing this works with named variants.

Reply:
sgtm